src/menu/menu.c

```diff
@@ -856,22 +970,200 @@ menu_open(struct menu *menu, int x, int y)
 	menu->server->input_mode = LAB_INPUT_STATE_MENU;
 }
 
+struct pipe_context {
+	struct server *server;
+	struct menuitem *item;
+	struct buf buf;
+	struct wl_event_source *event_read;
+	struct wl_event_source *event_timeout;
+	pid_t pid;
+	int pipe_fd;
+};
+
+static int
+handle_pipemenu_timeout(void *_ctx)
+{
+	struct pipe_context *ctx = _ctx;
+	kill(ctx->pid, SIGTERM);
+	return 0;
+}
+
+static int
+handle_pipemenu_readable(int fd, uint32_t mask, void *_ctx)
+{
+	struct pipe_context *ctx = _ctx;
+	char data[4096];
+	ssize_t size;
+
+	do {
+		/* leave space for terminating NULL byte */
+		size = read(fd, data, sizeof(data) - 1);
+	} while (size == -1 && errno == EINTR);
+
+	if (size == -1) {
+		perror("Failed to read from pipemenu");
+		wlr_log(WLR_ERROR, "Failed to read data for pipe menu");
+		goto clean_up;
+	}
+	wlr_log(WLR_INFO, "read %li bytes of data", size);
+	if (size) {
+		data[size] = '\0';
+		buf_add(&ctx->buf, data);
+		return 0;
+	}
```

Comment:
Should we limit the buffer size (`ctx->buf.len`) to something like 1 MB and when reached kill the script? Should help against pipemenu scripts accidentally filling up all the ram.

Reply:
Yes. Agree. 
Suspect 100k would be plenty!